File: android/src/main/kotlin/com/cloudwebrtc/webrtc/proxy/MediaStreamTrackProxy.kt
package com.cloudwebrtc.webrtc.proxy

import android.util.Log
import com.cloudwebrtc.webrtc.TrackRepository
import com.cloudwebrtc.webrtc.model.MediaKind
import com.cloudwebrtc.webrtc.model.MediaStreamTrackState
import org.webrtc.MediaStreamTrack

/**
 * Wrapper around [MediaStreamTrack].
 *
 * @param track underlying [MediaStreamTrack].
 * @property deviceId unique ID of device on which this [MediaStreamTrackProxy] is based.
 * @property source [MediaTrackSource] from which this [MediaStreamTrackProxy] is created.
 * `null` for [MediaStreamTrackProxy]s received from the remote side.
 */
class MediaStreamTrackProxy(
    track: MediaStreamTrack,
    private val deviceId: String = "remote",
    private val source: MediaTrackSource? = null
) :
    IWebRTCProxy<MediaStreamTrack> {
    /**
     * Actual underlying [MediaStreamTrack].
     */
    override var obj: MediaStreamTrack = track

    /**
     * ID of underlying [MediaStreamTrack].
     */
    private val id = track.id()

    /**
     * Subsribers for the [onStop] callback.
     *
     * Will be called once on [stop] call.
     */
    private var onStopSubscribers: MutableList<() -> Unit> = mutableListOf()

    /**
     * Indicates that this [stop] was called on this [MediaStreamTrackProxy].
     */
    private var isStopped: Boolean = false

    init {
        TrackRepository.addTrack(id, this)
    }

    override fun syncWithObject() {}

    /**
     * Returns ID of the underlying [MediaStreamTrack].
     *
     * @return ID of the underlying [MediaStreamTrack].
     */
    fun id(): String {
        return id
    }

    /**
     * @return [MediaKind] of the underlying [MediaStreamTrack].
     */
    fun kind(): MediaKind {
        return when (obj.kind()) {
            MediaStreamTrack.VIDEO_TRACK_KIND -> MediaKind.Video
            MediaStreamTrack.AUDIO_TRACK_KIND -> MediaKind.Audio
            else -> throw Exception("LibWebRTC provided unknown MediaKind value")
        }
    }

    /**
     * @return unique device ID of the underlying [MediaStreamTrack].
     */
    fun deviceId(): String {
        return deviceId
    }

    /**
     * Creates new [MediaStreamTrackProxy] based on the same [MediaTrackSource] as
     * this [MediaStreamTrackProxy].
     *
     * Can be called only on local [MediaStreamTrackProxy]s.
     *
     * @throws Exception if called on remote [MediaStreamTrackProxy].
     * @return created [MediaStreamTrackProxy].
     */
    fun clone(): MediaStreamTrackProxy {
        if (this.source == null) {
            throw Exception("Remote MediaStreamTracks can't be cloned")
        } else {
            return source.newTrack()
        }
    }

    /**
     * Stops this [MediaStreamTrackProxy].
     *
     * Media source will be disposed only if there is no another [MediaStreamTrackProxy]
     * depending on [MediaTrackSource].
     */
    fun stop() {
        if (!isStopped) {
            isStopped = true
            onStopSubscribers.forEach { sub -> sub() }
        } else {
            Log.w("FlutterWebRTC", "Double stop detected [deviceId: $deviceId]!")
        }
    }

    /**
     * @return [MediaStreamTrackState] of the underlying [MediaStreamTrack].
     */
    fun state(): MediaStreamTrackState {
        return MediaStreamTrackState.fromWebRtcState(obj.state())
    }

    /**
     * Sets enabled state of the underlying [MediaStreamTrack].
     *
     * @param enabled state which will be set to the underlying [MediaStreamTrack].
     */
    fun setEnabled(enabled: Boolean) {
        obj.setEnabled(enabled)
    }

    /**
     * Subscribes to the [stop] event of this [MediaStreamTrackProxy].
     *
     * This callback guaranteed to be called only once.
     */
    fun onStop(f: () -> Unit) {
        onStopSubscribers.add(f)
    }
}

File: ios/Classes/proxy/MediaStreamTrackProxy.swift
import WebRTC

/// Wrapper around a `MediaStreamTrack`.
class MediaStreamTrackProxy: Equatable {
  /// Indicates that this `stop` was called on this `MediaStreamTrackProxy`.
  private var isStopped = false

  /// Device ID of this `MediaStreamTrackProxy`.
  private var deviceId = "remote"

  /// Source of this `MediaStreamTrackProxy`.
  private var source: MediaTrackSource?

  /// Underlying `RTCMediaStreamTrack`.
  private var track: RTCMediaStreamTrack

  /// Subscribers for `onStop` callback of this `MediaStreamTrackProxy`.
  private var onStopSubscribers: [() -> Void] = []

  /// Subscribers for `onEnded` callback of this `MediaStreamTrackProxy`.
  private var onEndedSubscribers: [() -> Void] = []

  /// Creates a new `MediaStreamTrackProxy` based on the provided data.
  init(track: RTCMediaStreamTrack, deviceId: String?, source: MediaTrackSource?) {
    self.source = source
    if deviceId != nil {
      self.deviceId = deviceId!
    }
    self.track = track
    MediaStreamTrackStore.tracks[track.trackId] = self
  }

  /// Compares two `MediaStreamTrackProxy`s based on underlying `RTCMediaStreamTrack`s.
  static func == (lhs: MediaStreamTrackProxy, rhs: MediaStreamTrackProxy) -> Bool {
    lhs.track == rhs.track
  }

  /// Adds `RTCVideoRenderer` for the underlying `RTCMediaStreamTrack`.
  func addRenderer(renderer: RTCVideoRenderer) {
    let videoTrack = self.track as! RTCVideoTrack
    videoTrack.add(renderer)
  }

  /// Removes `RTCVideoRenderer` for the underlying `RTCMediaStreamTrack`.
  func removeRenderer(renderer: RTCVideoRenderer) {
    let videoTrack = self.track as! RTCVideoTrack
    videoTrack.remove(renderer)
  }

  /// Returns ID of this track.
  func id() -> String {
    track.trackId
  }

  /// Returns kind of this track.
  func kind() -> MediaType {
    let kind = track.kind
    switch kind {
    case "audio":
      return MediaType.audio
    case "video":
      return MediaType.video
    default:
      abort()
    }
  }

  /// Returns device ID of this track.
  func getDeviceId() -> String {
    self.deviceId
  }

  /// Returns forked `MediaStreamTrackProxy` based on the same source as this track.
  func fork() throws -> MediaStreamTrackProxy {
    if self.source == nil {
      throw MediaStreamTrackException.remoteTrackCantBeCloned
    } else {
      return source!.newTrack()
    }
  }

  /// Stops this track.
  ///
  /// Source will be stopped and disposed when all it's tracks are stopped.
  func stop() {
    self.isStopped = true
    for cb in onStopSubscribers {
      cb()
    }
  }

  /// Returns current `readyState` of this track.
  func state() -> MediaStreamTrackState {
    let state = track.readyState
    switch state {
    case .live:
      return MediaStreamTrackState.live
    case .ended:
      return MediaStreamTrackState.ended
    default:
      abort()
    }
  }

  /// Sets `enabled` state of this track.
  func setEnabled(enabled: Bool) {
    self.track.isEnabled = enabled
  }

  /// Subscribes on `onStopped` callback of this track.
  func onStopped(cb: @escaping () -> Void) {
    self.onStopSubscribers.append(cb)
  }

  /// Notifies `RtpReceiverProxy` about its `MediaStreamTrackProxy` being removed from the receiver.
  func notifyEnded() {
    if self.track.readyState == .ended {
      for cb in self.onEndedSubscribers {
        cb()
      }
    }
  }

  /// Subscribes on `onEnded` callback of this track.
  func onEnded(cb: @escaping () -> Void) {
    self.onEndedSubscribers.append(cb)
  }

  /// Returns underlying `RTCMediaStreamTrack` of this proxy.
  func obj() -> RTCMediaStreamTrack {
    self.track
  }
}
